ambari-server/src/main/java/org/apache/ambari/server/stack/ServiceModule.java

```diff
@@ -209,6 +210,10 @@ public void resolveInternal(
       ServiceInfo.ServiceAdvisorType serviceAdvisorType = parent.getServiceAdvisorType();
       serviceInfo.setServiceAdvisorType(serviceAdvisorType == null ? ServiceInfo.ServiceAdvisorType.PYTHON : serviceAdvisorType);
     }
+    if (serviceInfo.getCategory() == null) {
+      ServiceCategory category = parent.getCategory();
+      serviceInfo.setCategory(category == null ? ServiceCategory.LEGACY : category);
```

Comment:
Hi Dmitry,
I have a query. Inheritance will be present only for Legacy Stacks and hence services. So, when we say ServiceCategory category = parent.getCategory();, category will be null for new mpacks. In which case we will use the category from current module's metainfo.xml. Considering we have inserted a category to all the existing metainfo.xml files, what is the scenario for Category = LEGACY? Is it just for the very first version of a stack? Even for that, it should take its category from metainfo.xml instead of terming it LEGACY, correct?

Reply:
I haven't completely understood the issue here. 

In case the service is not inherited from other service this code will never be reached, instead the one from StackModule will be executed: https://github.com/d0zen1/ambari/blob/137d82ebfb8db273706c23a352803a8ed671a09b/ambari-server/src/main/java/org/apache/ambari/server/stack/StackModule.java#L356-L364

The general logic follows the following priority rules, given 1 is the highest priority:
1. Current service category (from service metainfo.xml if defined)
2. Parent service category (from parent service metainfo.xml if defined)
3. Default service category (LEGACY)